### ISSUE TITLE
Introduce a new source plugin - `SourceTransform`

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#381]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/381)
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Apr 30, 2018, 16:22

Aim of this issue is to provide a new source plugin `SourceTransform` that will be useful to handle sources of languages that come with their own package manager. It should provide a way to handle external dependencies of such sources using their package manager, for example, `dep` for GoLang or `cargo` for Rust.

This was originally discussed on the mailing list under [this thread](https://mail.gnome.org/archives/buildstream-list/2018-April/msg00034.html) (follow replies to this thread). I'm creating an issue here to track the implementation and to have discussions about the technical issues (if needed).

To summarize, we want to add a new source plugin - `SourceTransform` - that will be a kind of source but a bit special (thanks to [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) for the idea):

- It can only appear after other sources (similar to `patch`)
- It will have an additional directory in context, which will contain the result of staging all previous sources.
- It can only be tracked if all other sources before it have been fetched (not just tracked).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on May 1, 2018, 08:59

Hi, is there any timeline for this? For our project we are going for this approach for now: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/merge_requests/190

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @awacheux]](https://gitlab.com/awacheux) on May 9, 2018, 12:07

I'm going to have a look on this.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @gokcennurlu]](https://gitlab.com/gokcennurlu) on Jun 18, 2018, 15:05

mentioned in merge request !505

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 13, 2018, 09:50

assigned to [[Gitlab user @gokcennurlu]](https://gitlab.com/gokcennurlu)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 25, 2018, 19:53

mentioned in commit 7da1aa3f9fee4b2a0cdec18359f3e55d487c6afb

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 25, 2018, 19:56

mentioned in merge request !568

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 25, 2018, 19:59

mentioned in commit 51532b3358ca1fe01122539d2d9e1a2dada9f97a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jul 26, 2018, 07:42

I like the new approach of extending the base `Source` API in !568, since this seems to not add too much, it's nice to just extend `Source`.

Orthogonally to this, I'd like to reiterate my concern raised in the previous MR: https://gitlab.com/BuildStream/buildstream/merge_requests/505#note_85196540

My worry here is that we land this core functionality in BuildStream proper, and third parties develop plugins which use this on their own without submitting those plugins upstream.

This would result in the core functionality being available only to those who want to write a complex plugin, and not benefiting the wider user base, that would be very bad. We need to strive to have coverage of the more popular use cases of this, such as `cargo` (rust) and `godep` (go) available in the core BuildStream set of plugins.

We should make inclusion of this feature in `1.4` contingent on landing at least one `Source` plugin which provides this functionality to users.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 27, 2018, 14:01

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) that makes sense to me.

We do have a version of `godep` that sort of works but we are a bit hesitant to land that in BuildStream core as we fear that its API might not be stable just yet. It might make sense to let it develop in bst-external while its API matures and move it back here once people think it's stable enough.

But, we are also planning to work on a `pip` plugin for python, which I think should be easier to maintain as most of us are more familiar with Python (compared to Go). We are working towards adding that soon.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 28, 2018, 21:18

mentioned in commit a9138f20bccbef38ae66a973c668c14ad9b36c5c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 30, 2018, 11:29

mentioned in commit ecb7e0a72baf6e8f4452a1eaa97937ac0eeb9a64

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 30, 2018, 11:56

mentioned in commit 39c37a645b255466afd55b065dae79af284aaad3

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 30, 2018, 17:17

mentioned in commit f5a2267e368273bb0d88f1ff0bb80b2c847ba6a2

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 2, 2018, 15:29

mentioned in commit 08d77356b437e923156a9b382bf56cfdaa9983be

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 2, 2018, 15:30

mentioned in commit a0b2d7ca2185f8f793c2550c6ca9c5dc12df0ebc

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 2, 2018, 17:18

mentioned in commit e5d4aad09f270b97bb9b4a96a0d072632a7b764f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 7, 2018, 16:18

mentioned in commit c90ed1e4a5e4d11a4274b2d8bcc75802327fc3e0

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 7, 2018, 19:01

mentioned in commit f85eb5007662e18eee42e4c129907b158378fcf7

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 10, 2018, 12:31

mentioned in commit a86bf81f9ccf7586351aa0e295c34eae6bd09129

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 10, 2018, 18:51

mentioned in commit daaf301723c3b0d6b8eed2e80e44cd634385070a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 13, 2018, 11:40

mentioned in commit 94eda15ee3b8a390ce01fc89b5be3430a12db44a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 11:23

mentioned in commit c55239bac746e3e8a513b86c5d447d9303934a35

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 12:51

mentioned in commit 89ba99c992da3aa801a3223bb08c734d27c67b15

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 13:38

mentioned in commit 673531c89b1d60337345b377caacae0f7190e1a9

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 16:53

mentioned in commit 74fc15dad529eca6b4d7ca02673218e83a8cfce3

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 18:29

mentioned in commit 9d2442b2794d7a531f50b3f1f9a0c6e4236bd9be

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 21:21

closed via merge request !568

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Aug 15, 2018, 21:21

mentioned in commit d4706096abffe54855269f16261aba06b5fa09b6

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 15, 2019, 15:22

mentioned in commit 567f597030a78558209660365bf764c3ffd229ef

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 15, 2019, 15:30

mentioned in commit f29edace2183976015310dbbf96f2c138d72381f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jul 16, 2019, 06:05

mentioned in commit f8cf48c8a5819740f03b1aaf3f45b4ab1571bc15
